Userland/DevTools/UserspaceEmulator/SoftFPU.h

```diff
@@ -0,0 +1,665 @@
+    {
+        reportln("Stack Overflow");
+        set_c1(1);
+        fpu_set_exception(FPU_Exception::StackFault);
+    }
+
+    void fpu_set_stack_underflow()
+    {
+        reportln("Stack Underflow");
+        set_c1(0);
+        fpu_set_exception(FPU_Exception::StackFault);
+    }
+
+    constexpr FPU_Tag fpu_get_tag_absolute(u8 index) const
+    {
+        switch (index) {
+        case 0:
+            return FPU_Tag(m_fpu_status_0);
+        case 1:
+            return FPU_Tag(m_fpu_status_1);
+        case 2:
+            return FPU_Tag(m_fpu_status_2);
+        case 3:
+            return FPU_Tag(m_fpu_status_3);
+        case 4:
+            return FPU_Tag(m_fpu_status_4);
+        case 5:
+            return FPU_Tag(m_fpu_status_5);
+        case 6:
+            return FPU_Tag(m_fpu_status_6);
+        case 7:
+            return FPU_Tag(m_fpu_status_7);
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    constexpr FPU_Tag fpu_get_tag(u8 index) const
+    {
+        VERIFY(index < 8);
+        return fpu_get_tag_absolute((m_fpu_stack_top + index) % 8);
+    }
+
+    void fpu_set_tag_absolute(u8 index, FPU_Tag tag)
+    {
+        switch (index) {
+        case 0:
+            m_fpu_status_0 = (u8)tag;
+            break;
+        case 1:
+            m_fpu_status_1 = (u8)tag;
+            break;
+        case 2:
+            m_fpu_status_2 = (u8)tag;
+            break;
+        case 3:
+            m_fpu_status_3 = (u8)tag;
+            break;
+        case 4:
+            m_fpu_status_4 = (u8)tag;
+            break;
+        case 5:
+            m_fpu_status_5 = (u8)tag;
+            break;
+        case 6:
+            m_fpu_status_6 = (u8)tag;
+            break;
+        case 7:
+            m_fpu_status_7 = (u8)tag;
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    void fpu_set_tag(u8 index, FPU_Tag tag)
+    {
+        VERIFY(index < 8);
+        fpu_set_tag_absolute((m_fpu_stack_top + index) % 8, tag);
+    }
+
+    void set_tag_from_value_absolute(u8 index, long double val)
+    {
+        // (FP_NAN, FP_INFINITE, FP_ZERO, FP_SUBNORMAL, FP_ZERO,
+        switch (fpclassify(val)) {
+        case FP_ZERO:
+            fpu_set_tag_absolute(index, FPU_Tag::Zero);
+            break;
+        case FP_NAN:
+        case FP_INFINITE:
+        case FP_SUBNORMAL:
+            fpu_set_tag_absolute(index, FPU_Tag::Special);
+            break;
+        case FP_NORMAL:
+            fpu_set_tag_absolute(index, FPU_Tag::Valid);
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    void set_tag_from_value(u8 index, long double val)
+    {
+        set_tag_from_value_absolute((m_fpu_stack_top + index) % 8, val);
+    }
+
+    bool fpu_isnan(u8 index) const
+    {
+        return (fpu_get_tag(index) == FPU_Tag::Special) && isnan(fpu_get(index));
+    }
+
+    bool fpu_is_set(u8 index) const
```

Comment:
`fpu_is_set` sounds a bit awkward. How about `has_value_at`?

Reply:
It can have a value even if it isn't set though...